File: yggdrasil/examples/formatted_io6/src/formatted_io6_modelB.c
#include <stdio.h>
// Include methods for input/output channels
#include "YggInterface.h"

int main(int argc, char *argv[]) {
  // Initialize input/output channels
  yggObjInput_t in_channel = yggObjInput("inputB");
  yggObjOutput_t out_channel = yggObjOutput("outputB");

  // Declare resulting variables and create buffer for received message
  int flag = 1;
  obj_t p = init_obj();

  // Loop until there is no longer input or the queues are closed
  while (flag >= 0) {
  
    // Receive input from input channel
    // If there is an error, the flag will be negative
    // Otherwise, it is the size of the received message
    flag = yggRecv(in_channel, &p);
    if (flag < 0) {
      printf("Model B: No more input.\n");
      break;
    }

    // Print received message
    printf("Model B: (%d verts, %d faces)\n", p.nvert, p.nface);
    display_obj_indent(p, "  ");

    // Send output to output channel
    // If there is an error, the flag will be negative
    flag = yggSend(out_channel, p);
    if (flag < 0) {
      printf("Model B: Error sending output.\n");
      break;
    }

  }
  
  // Free dynamically allocated obj structure
  free_obj(&p);
  
  return 0;
}

File: yggdrasil/examples/formatted_io6/src/formatted_io6_modelB.f90
program main
  ! Include methods for input/output channels
  use fygg

  ! Declare resulting variables and create buffer for received message
  logical :: flag = .true.
  type(yggcomm) :: in_channel, out_channel
  type(yggobj) :: p
  p = init_obj()

  ! initialize input/output channels
  in_channel = ygg_obj_input("inputB")
  out_channel = ygg_obj_output("outputB")

  ! Loop until there is no longer input or the queues are closed
  do while (flag)

     ! Receive input from input channel
     ! If there is an error, the flag will be negative
     ! Otherwise, it is the number of variables filled
     flag = ygg_recv_var(in_channel, yggarg(p))
     if (.not.flag) then
        print *, "Model B: No more input."
        exit
     end if

     ! Print received message
     print *, "Model B: (", &
          nelements_obj(p, "v"), " verts, ", &
          nelements_obj(p, "f"), " faces)"
     call display_obj_indent(p, "  ")

     ! Send output to output channel
     ! If there is an error, the flag will be negative
     flag = ygg_send_var(out_channel, yggarg(p))
     if (.not.flag) then
        print *, "Model B: Error sending output."
        exit
     end if

  end do

end program main
